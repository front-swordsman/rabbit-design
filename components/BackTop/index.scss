$backtop: 'snake-design-backtop';

.#{$backtop} {
  &-default {
    position: fixed;
    right: 100px;
    bottom: 50px;
    width: 40px;
    height: 40px;
    line-height: 40px;
    border-radius: 50%;
    background: rgba(0, 0, 0, 0.45);
    text-align: center;
  }

  &-hide {
    display: none;
  }
}
