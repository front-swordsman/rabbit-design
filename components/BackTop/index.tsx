import * as React from 'react'
import { BacktopProps } from 'types/backtop'
import cx from 'classnames'
import Icon from '../Icon'
import { throttle } from '../utils/tool'
import './index.scss'

const { useState, useEffect } = React

const defaultProps = {
  prefixCls: 'snake-design-backtop'
}

const BackTop = (userProps: BacktopProps, ref: any) => {
  const props = {
    ...userProps,
    ...defaultProps
  }

  const { prefixCls, children } = props

  const [show, setShow] = useState(false)

  const scrollLogic = () => {
    if (window.scrollY >= 400) {
      setShow(true)
    } else {
      setShow(false)
    }
  }

  useEffect(() => {
    const scroll = throttle(scrollLogic, 20)
    window.addEventListener('scroll', scroll)
    return () => {
      window.removeEventListener('scroll', scroll)
    }
  }, [])

  const backTopFn = () => {
    window.scrollTo({
      top: 0,
      behavior: 'smooth'
    })
  }

  return (
    <div
      ref={ref}
      className={cx(`${prefixCls}-default`, {
        [`${prefixCls}-hide`]: !show
      })}
      onClick={backTopFn}
    >
      {children ? children : <Icon type="totop" color="#fff" />}
    </div>
  )
}

export default React.forwardRef(BackTop)
