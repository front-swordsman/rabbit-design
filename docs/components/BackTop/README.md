---
imports: import BackTopDemoCode from '!raw-loader!./simpleBackTop.tsx';
  import BackTopDemo from './simpleBackTop.tsx';
  import CustomBackTopCode from '!raw-loader!./customBackTop.tsx';
  import CustomBackTopDemo from './customBackTop.tsx';
---

# BackTop 回到顶部

返回页面顶部的操作按钮。

### 何时使用

* 当页面内容区域比较长时；
* 当用户需要频繁返回顶部查看相关内容时；

### 简单示例

:::demo
<Block code={BackTopDemoCode} des="最简单的用法">
<BackTopDemo />
</Block>
:::

:::demo
<Block code={CustomBackTopCode} des="自定义样式">
<CustomBackTopDemo />
</Block>
:::

## Props

> 有默认样式，距离底部 50px，可覆盖。

| 参数             | 说明                               | 类型            | 默认值 |
| :--------------- | :--------------------------------- | :-------------- | :----- |
| visibilityHeight | 滚动高度达到此参数值才出现 BackTop | number          | 400    |
| children         | 传入子节点                         | React.ReactNode | -      |
| className        | 样式                               | string          | -      |
| style            | 样式                               | object          | -      |
