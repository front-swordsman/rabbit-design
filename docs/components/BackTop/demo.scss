.snake-back-top-inner {
  height: 40px;
  width: 40px;
  line-height: 40px;
  border-radius: 4px;
  background-color: #01C1B2;
  color: #fff;
  text-align: center;
  font-size: 20px;
}

#components-back-top-demo-custom .snake-design-backtop-default {
  bottom: 100px;
}