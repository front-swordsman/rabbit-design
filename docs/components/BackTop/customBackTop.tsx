import * as React from 'react'
import BackTop from 'components/BackTop'
import './demo.scss'

function CustomBackTop() {
  return (
    <>
      <p>
        Scroll down to see the bottom-right <i style={{ color: '#01C1B2' }}>green</i> button.
      </p>
      <div id="components-back-top-demo-custom">
        <BackTop>
          <div className="snake-back-top-inner">UP</div>
        </BackTop>
      </div>
    </>
  )
}

export default CustomBackTop
