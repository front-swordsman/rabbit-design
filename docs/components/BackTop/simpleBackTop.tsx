import * as React from 'react'
import BackTop from 'components/BackTop'
import './demo.scss'

function SimpleBackTop() {
  return (
    <>
      <p>Scroll down to see the bottom-right gray button.</p>
      <BackTop />
    </>
  )
}

export default SimpleBackTop
