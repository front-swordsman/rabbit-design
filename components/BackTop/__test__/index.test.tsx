import * as React from 'react'
import { render, fireEvent } from 'react-testing-library'
import BackTop from '../index'

describe('BackTop Test', () => {
  it('snapshot without children', () => {
    const { container } = render(<BackTop />)
    expect(container).toMatchSnapshot()
  })

  it('snapshot with children', () => {
    const { container } = render(
      <BackTop>
        <div>回到顶部</div>
      </BackTop>
    )
    expect(container).toMatchSnapshot()
  })

  it('click backTop back to the top', () => {
    const { getByText } = render(<BackTop>回到顶部</BackTop>)
    window.scrollTo(0, 500)
    const back = getByText(/回到顶部/i) as HTMLElement
    fireEvent.click(back)
    expect(document.documentElement.scrollTop).toBe(0)
  })
})
