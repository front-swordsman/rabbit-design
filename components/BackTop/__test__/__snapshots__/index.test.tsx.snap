// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BackTop Test snapshot with children 1`] = `
<div>
  <div
    class="snake-design-backtop-default snake-design-backtop-hide"
  >
    <div>
      回到顶部
    </div>
  </div>
</div>
`;

exports[`BackTop Test snapshot without children 1`] = `
<div>
  <div
    class="snake-design-backtop-default snake-design-backtop-hide"
  >
    <svg
      class="snake-icon snake-icon-totop"
      style="color: rgb(255, 255, 255);"
    >
      <use
        xlink:href="#icon-totop"
      />
    </svg>
  </div>
</div>
`;
